### PR TITLE
Issue 81 improved cmake

### PR DESCRIPTION
**Pull request by ChristianVonMach**
Attempt to improve the general cmake structure before adding the unit tests

### CODE REVIEW

File: compile.sh [hunk 1]
```diff
@@ -47,14 +46,18 @@ Executables:
...
   ${YELLOW}--cc${NC}
   ${YELLOW}--computecompatibility${NC}            \t set compute compability of the GPU (30|35|50|60|70|75)
-   ${YELLOW}-d${NC}
-   ${YELLOW}--debugmode${NC}                      \t set debug flag for build type (default: ${BUILDTYPE})
+  ${YELLOW}-d${NC}
+  ${YELLOW}--debugmode${NC}                       \t set debug flag for build type (default: ${BUILDTYPE})
+  ${YELLOW}-t${NC}
+  ${YELLOW}--test${NC}                            \t enables the test target for make 
+  ${YELLOW}-u${NC}
+  ${YELLOW}--utest${NC}                           \t enables the unit test target for make
```
Comment (LinhWuerzburger) [inline compile.sh:58]: write it out: unittest instead of utest

File: compile.sh [hunk 2]
```diff
@@ -47,14 +46,18 @@ Executables:
   ${YELLOW}--artss_gpu_benchmark${NC}               \t Executable artss_gpu_benchmark
 
 Other:
-   ${YELLOW}-c${NC}
```
Comment (LinhWuerzburger) [inline compile.sh:50]: the blank is on purpose to align the text

File: compile.sh [hunk 3]
```diff
@@ -142,12 +146,13 @@ do
...
       ;;
     -m|--multicore|--artss_multicore_cpu)
-      COMPILE="$COMPILE artss_multicore_cpu"
+      CMAKE_FLAGS="$CMAKE_FLAGS -DARTSS_TARGET_ARCH=MUTICORE" 
       GPU=0
       shift
       ;;
     --mb|--multicore_benchmark|--artss_multicore_cpu_benchmark)
-      COMPILE="$COMPILE artss_multicore_cpu_benchmark"
+      CMAKE_FLAGS="$CMAKE_FLAGS -DARTSS_TARGET_ARCH=MUTICORE"
```
Comment (LinhWuerzburger) [inline compile.sh:154]: missing L in multicore

File: compile.sh [hunk 4]
```diff
@@ -161,11 +166,20 @@ do
...
+      CMAKE_FLAGS="$CMAKE_FLAGS -DARTSS_TARGET_ARCH=SERIAL" 
+      CMAKE_FLAGS="$CMAKE_FLAGS -DARTSS_BENCHMARK_FLAG=ON"
+      shift
+      ;;
+    -t|--test)
+      CMAKE_FLAGS="$CMAKE_FLAGS -DARTSS_TESTS_FLAG=ON" 
+      shift
+      ;;
+    -u|--utest)
+      CMAKE_FLAGS="$CMAKE_FLAGS -DARTSS_UTEST_FLAG=ON"
```
Comment (LinhWuerzburger) [inline compile.sh:182]: also unittest instead of utest

File: compile.sh [hunk 5]
```diff
@@ -268,15 +282,15 @@ fi
...
+  cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=1 -DCMAKE_BUILD_TYPE=${BUILDTYPE} -DCMAKE_C_COMPILER=${CCOMPILER} -DCMAKE_CXX_COMPILER=${CXXCOMPILER} ${CMAKE_FLAGS} ..
 else
-  cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=1 -DCMAKE_BUILD_TYPE=${BUILDTYPE} -DCMAKE_C_COMPILER=${CCOMPILER} -DCMAKE_CXX_COMPILER=${CXXCOMPILER} -DGPU_CC=${GPU_CC} -DCUDA_VERSION=${CUDA_VERSION} ..
+  cmake -DCMAKE_EXPORT_COMPILE_COMMANDS=1 -DCMAKE_BUILD_TYPE=${BUILDTYPE} -DCMAKE_C_COMPILER=${CCOMPILER} -DCMAKE_CXX_COMPILER=${CXXCOMPILER} -DGPU_CC=${GPU_CC} -DCUDA_VERSION=${CUDA_VERSION} ${CMAKE_FLAGS} ..
 fi
 
 if [ "$PROCS" -le 0 ]
 then
-  make $COMPILE
+  make artss
```
Comment (LinhWuerzburger) [inline compile.sh:292]: what happens if you just use make?

**Review by LinhWuerzburger — CHANGES_REQUESTED** [review]
just wondering about consistency: in a few cmake files there is a new line at the ende of the file and there is also a file which has a new line inside the brackets

and of course the in-line comments have to be discussed

File: compile.sh (hunk 5 above)
Comment (ChristianVonMach) [reply compile.sh:292]: it will also build implicit artss as default target. only reason is to be more explicit.

Comment (LinhWuerzburger) [reply compile.sh:292]: okay

File: compile.sh (hunk 1 above)
Comment (ChristianVonMach) [reply compile.sh]: done in c51c52b

File: compile.sh (hunk 4 above)
Comment (ChristianVonMach) [reply compile.sh:182]: done in c51c52b

File: compile.sh (hunk 3 above)
Comment (ChristianVonMach) [reply compile.sh:154]: done in 57e7ac8

File: compile.sh (hunk 2 above)
Comment (ChristianVonMach) [reply compile.sh:50]: done in 13f32dd

**Comment by ChristianVonMach** [comment]
> just wondering about consistency: in a few cmake files there is a new line at the ende of the file and there is also a file which has a new line inside the brackets
> 
> and of course the in-line comments have to be discussed

fixed inconsistency in eaf8d33 and newline in 17d6b69

File: compile.sh (hunk 2 above)
Comment (LinhWuerzburger) [reply compile.sh:50]: if you change it from blank ${YELLOW}-c to ${YELLOW} blank -c, you should do it thoroughly. or change it back.

File: CMakeLists.txt [hunk 6]
```diff
@@ -8,12 +8,17 @@
...
-cmake_minimum_required(VERSION 2.8  FATAL_ERROR)
+cmake_minimum_required(VERSION 2.8 FATAL_ERROR)
 
 project(ARTSS)
 
 set(CMAKE_COLOR_MAKEFILE ON)
 
+option(ARTSS_BENCHMARK_FLAG "Flag if BENCHMARKING is enabled" OFF)
+option(ARTSS_TESTS_FLAG "Flag if testing is enabled" OFF)
+option(ARTSS_UTEST_FLAG "Flag if unit testing is enabled" OFF)
```
Comment (LinhWuerzburger) [inline CMakeLists.txt:19]: here as well with unittest

**Comment by LinhWuerzburger** [comment]
image
I like the new information, but maybe just one new line after the platform section. And I was thinking about putting the four lines regarding the FLAGS to the platform section as well. To structure it a bit more

File: CMakeLists.txt (hunk 6 above)
Comment (ChristianVonMach) [reply CMakeLists.txt]: done in 9ce285e

**Comment by ChristianVonMach** [comment]
> image
> I like the new information, but maybe just one new line after the platform section. And I was thinking about putting the four lines regarding the FLAGS to the platform section as well. To structure it a bit more

fixed in c182035. New ouput looks like this.
image

**Comment by LinhWuerzburger** [comment]
Build Type: Release/Debug is missing when the benchmarking flag is on? just tested with serial version so far

**Comment by LinhWuerzburger** [comment]
I like the attempt to simplify the cmake file, but it doesn't seem practical to me. The time to create multiple versions or to rebuild a version for minor changes is now so much longer than before. With this I come to the conclusion: I prefer an overloaded/messy cmake file over longer compilation times.

**Pull request closed without merging**